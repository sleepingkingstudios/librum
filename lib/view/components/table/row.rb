# frozen_string_literal: true

require 'view/components/table'

module View::Components
  # Component for rendering a table row.
  class Table::Row < ViewComponent::Base
    # @param columns [Array<View::Components::DataField::FieldDefinition>] the
    #   columns used to render the table.
    # @param data [Hash{String=>Object}] the data object for the row.
    # @param cell_component [ViewComponent::Base] the component to render each
    #   table cell. Defaults to View::Components::Table::Cell.
    # @param options [Hash{Symbol=>Object}] additional options to pass to the
    #   rendered row.
    def initialize(columns:, data:, cell_component: nil, **options)
      super()

      @columns        = columns
      @data           = data
      @cell_component = cell_component || View::Components::Table::Cell
      @options        = options
    end

    # @return [ViewComponent::Base] the component to render each table cell.
    attr_reader :cell_component

    # @return [View::Components::DataField::FieldDefinition] the columns used to
    #   render the table.
    attr_reader :columns

    # @return [Array<Hash{String=>Object}>] the table data to render.
    attr_reader :data

    # @return [Hash{Symbol=>Object}] additional options to pass to the rendered
    #   cell.
    attr_reader :options

    private

    def build_cell(column:)
      cell_component.new(column: column, data: data, **options)
    end

    def render_cell(column:)
      render(build_cell(column: column))
    end
  end
end
