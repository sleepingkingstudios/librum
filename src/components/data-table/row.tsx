import * as React from 'react';

import { DataTableCell } from './cell';
import type { DataTableColumn } from './types';

export const DataTableRow = ({
  columns,
  data,
}: {
  columns: DataTableColumn[],
  data: Record<string, unknown>,
}): JSX.Element => {
  return (
    <tr>
      {
        columns.map(
          (column: DataTableColumn): JSX.Element => (
            <DataTableCell
              column={column}
              data={data}
              key={column.name}
            />
          )
        )
      }
    </tr>
  );
};
