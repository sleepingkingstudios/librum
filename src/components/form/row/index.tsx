import * as React from 'react';

import { joinClassNames } from '@utils/react-utils';
import type {
  UpToSixColumns,
  UpToTwelveColumns,
} from '../../types';

interface IFormRowProps {
  children: React.ReactNode;
  className?: string;
  cols?: UpToTwelveColumns;
  colsSm?: UpToSixColumns;
  reverse?: true;
}

interface IGridClassName {
  cols?: UpToTwelveColumns,
  colsSm?: UpToSixColumns,
}

const gridClassName = ({ cols, colsSm }: IGridClassName): string | null => {
  if (cols === null) { return null; }

  const largeColumns = cols;
  const smallColumns = colsSm ? colsSm : Math.round(cols / 2);

  return joinClassNames(
    'gap-3 sm:grid',
    `sm:grid-cols-${smallColumns}`,
    `lg:grid-cols-${largeColumns}`,
  );
};

export const FormRow = ({
  children,
  className = null,
  cols = null,
  colsSm = null,
  reverse = null,
}: IFormRowProps): JSX.Element => {
  const joinedClassName = joinClassNames(
    gridClassName({ cols, colsSm }),
    className,
  );
  const style: React.CSSProperties = reverse ? { direction: 'rtl' } : {};

  return (
    <div className={joinedClassName} style={style}>{ children }</div>
  );
}
