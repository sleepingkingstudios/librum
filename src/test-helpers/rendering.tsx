import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
