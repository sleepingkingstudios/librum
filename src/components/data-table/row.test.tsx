import * as React from 'react';

import '@testing-library/jest-dom';
import { render } from '@test-helpers/rendering';

import { DataTableRow } from './row';
import type { DataTableColumn } from './types';

type SpaceMission = {
  deltaV: string,
  launchDate?: string,
  missionName?: string,
  missionObjective: string,
  ordinal: string,
  program: string,
};

const RowWrapper = ({
  children,
}: {
  children: React.ReactNode,
}): JSX.Element => (<table><tbody>{ children }</tbody></table>);

describe('<DataTableRow />', () => {
  const Actions = (): JSX.Element => (
    <span className="data-table-actions">Launch</span>
  );
  const columns: DataTableColumn[] = [
    {
      name: 'missionName',
      value: ({ data }: { data: SpaceMission }) => (
        `${data.program} ${data.ordinal}`
      ),
    },
    {
      label: 'Objective',
      name: 'missionObjective',
    },
    {
      name: 'deltaV',
      numeric: true,
    },
    {
      align: 'center',
      default: 'TBD',
      name: 'launchDate',
    },
    {
      label: false,
      name: 'actions',
      value: Actions,
    },
  ];
  const data: SpaceMission = {
    deltaV: '1,000 km/s',
    missionObjective: '40 Eridani A',
    ordinal: 'VI',
    program: 'Hephaestus',
  };
  const expected = [
    'Hephaestus VI',
    '40 Eridani A',
    '1,000 km/s',
    'TBD',
    'Launch',
  ];

  it('should display the formatted data', () => {
    const { getAllByRole } = render(
      <DataTableRow columns={columns} data={data} />,
      { wrapper: RowWrapper },
    );
    const cells = getAllByRole('cell');

    expect(cells).toHaveLength(5);
    expect(cells.map(cell => cell.textContent)).toEqual(expected);
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(
      <DataTableRow columns={columns} data={data} />,
      { wrapper: RowWrapper },
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
