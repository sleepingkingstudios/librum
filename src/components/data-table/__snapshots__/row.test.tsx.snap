// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataTableRow /> should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Hephaestus VI
        </td>
        <td
          class="data-table-cell-mission-objective"
          data-label="Objective"
        >
          40 Eridani A
        </td>
        <td
          class="data-table-cell-delta-v text-right lining-nums tabular-nums"
          data-label="Delta V"
        >
          1,000 km/s
        </td>
        <td
          class="data-table-cell-launch-date text-center"
          data-label="Launch Date"
        >
          TBD
        </td>
        <td
          class="data-table-cell-actions"
          data-label="Actions"
        >
          <span
            class="data-table-actions"
          >
            Launch
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
